chore: Use client instead of APIReader + Add observedGeneration to status Conditions

michaelhtm opened the pull request:
Description of changes:
Currently in ACK, at the beginning of the Reconciler for each event, we 
get the resource from the k8s APIServer. This is an unecessary call, as 
the latest object that triggered the reconciler is already stored in the 
informers cache. 

With this change, we will be retrieving the resource from the cache 
instead of the APIServer.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

michaelhtm committed:
chore: Get resource from cache instead of APIServer

Currently in ACK, at the beginning of the Reconciler for each event, we
get the resource from the k8s APIServer. This is an unecessary call, as
the latest object that triggered the reconciler is already stored in the
informers cache.

With this change, we will be retrieving the resource from the cache
instead of the APIServer.

a-hilaly commented on pkg/runtime/adoption_reconciler.go:
```diff
@@ -625,7 +626,7 @@ func NewAdoptionReconcilerWithClient(
 			cache:     cache,
 			kc:        kc,
-			apiReader: apiReader,
+			ackResourceCache: ackResourceCache,
```
i think this can be gofmt-ed

a-hilaly commented on pkg/runtime/reconciler_test.go:
```diff
@@ -505,7 +505,7 @@ func TestReconcilerAdoptOrCreateResource_Adopt(t *testing.T) {
 	latest.On("Identifiers").Return(ids)
 	latest.On("Conditions").Return([]*ackv1alpha1.Condition{})
-		latest.On(
+	latest.On(
```
👍

a-hilaly reviewed (commented):
Dankeshoen!

michaelhtm committed:
gofmt all files

michaelhtm commented:
/test s3-controller-test

michaelhtm commented:
/test sagemaker-controller-test

knottnt commented on pkg/runtime/field_export_reconciler.go:
```diff
@@ -197,15 +198,15 @@ func (r *fieldExportReconciler) getFieldExport(
 	// It is okay to make direct call to k8s apiserver because we are only
 	// making single read call for complete reconciler loop.
 	// See following issue for more details:
 	// https://github.com/aws-controllers-k8s/community/issues/894
```
Is the problem described in this issue no longer relevant?

michaelhtm replied:
It's not related to a problem..it's mainly just an optimization.
I'll expand more on it in a community PR

michaelhtm committed:
Add ObservedGeneration to ACK Status Conditions

Observed Generation is a kubernetes standard that persists the latest
reconciled generation (metadata.generation) in the resource status
condition

michaelhtm committed:
Use controller-runtime Client instead of Cache

Using Controller-Runtime Client is safer in general, as it will attempt
to read objects from the APIServer if it was not cached in locally.

For example, when using resource references, or using ACK to reference
secrets, we may need to read these objects from the APIServer.

michaelhtm commented:
/retest

michaelhtm commented:
/close
